### PR TITLE
Implement all TODO coverage

### LINKED ISSUE #664: Implement all "TODO coverage"

### PR DESCRIPTION
**Pull request by aslesarenko**
Closes #664 
Also:
- work time estimations (somewhat conservative) for each `TODO HF` item.
- new test cases for all `TODO coverage` items
- fixed Option serialization in DataJsonEncoder

### CODE REVIEW

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala [hunk 1]
```diff
@@ -193,9 +196,17 @@ trait Interpreter extends ScorexLogging {
                     context: CTX,
                     env: ScriptEnv): (SigmaBoolean, Long) = {
     implicit val vs: SigmaValidationSettings = context.validationSettings
-    val prop = propositionFromErgoTree(ergoTree, context)
-    val (propTree, context2) = trySoftForkable[(BoolValue, CTX)](whenSoftFork = (TrueLeaf, context)) {
-      applyDeserializeContext(context, prop)
+
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:199]: move this to another PR, this change is not explained in PR description, also, it is not a good idea in general to mix changes in consensus-critical code with "todo coverage"

File: sigmastate/src/main/scala/sigmastate/serialization/DataJsonEncoder.scala [hunk 2]
```diff
@@ -46,7 +46,7 @@ object DataJsonEncoder {
   }
 
   private def encodeData[T <: SType](v: T#WrappedType, tpe: T): Json = tpe match {
-    case SUnit => Json.Null
+    case SUnit => Json.fromFields(mutable.WrappedArray.empty)
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/serialization/DataJsonEncoder.scala:49]: Seems like JSON format has been changed?  Why? Why not covered in PR description?

File: sigmastate/src/main/scala/sigmastate/serialization/DataJsonEncoder.scala [hunk 3]
```diff
@@ -95,7 +94,9 @@ object DataJsonEncoder {
     case tOpt: SOption[a] =>
       val opt = v.asInstanceOf[tOpt.WrappedType]
       if (opt.isDefined) {
-        encodeData(opt.get, tOpt.elemType)
+        // save the single value as an array with one item
+        val valueJson = encodeData(opt.get, tOpt.elemType)
+        Json.fromValues(Array(valueJson))
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/serialization/DataJsonEncoder.scala:99]: Seems like JSON format has been changed?  Why? Why not covered in PR description?

File: sigmastate/src/main/scala/sigmastate/serialization/DataSerializer.scala [hunk 4]
```diff
@@ -114,10 +114,7 @@ object DataSerializer {
         SigmaDsl.avlTree(AvlTreeData.serializer.parse(r))
       case tColl: SCollectionType[a] =>
         val len = r.getUShort()
-        if (tColl.elemType == SByte)
-          Colls.fromArray(r.getBytes(len))
-        else
-          deserializeColl(len, tColl.elemType, r)
+        deserializeColl(len, tColl.elemType, r)
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/serialization/DataSerializer.scala:117]: Please describe why this change is not breaking (better, do a test).

**Review by kushti — CHANGES_REQUESTED** [review]

File: sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala (hunk 1 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala:199]: rolled it back

File: sigmastate/src/main/scala/sigmastate/serialization/DataJsonEncoder.scala (hunk 2 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/serialization/DataJsonEncoder.scala:49]: Yes, this fixes a bug with option serialization. 
The tests are added to DataJsonEncoderSpecification.
PR description updated.

File: sigmastate/src/main/scala/sigmastate/serialization/DataSerializer.scala (hunk 4 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/serialization/DataSerializer.scala:117]: This is covered by test, so not tests are necessary. 
Actually, the coverage tool shows that `case SByte =>` in `deserializeColl` is never executed, because it is covered by `if (tColl.elemType == SByte)`.

So if you can compare the following fragments to see this is not breaking change.
```
if (tColl.elemType == SByte)
          Colls.fromArray(r.getBytes(len))
```
 and 
```
case SByte =>
        Colls.fromArray(r.getBytes(len)).asInstanceOf[Coll[T#WrappedType]]
```

File: sigmastate/src/main/scala/sigmastate/serialization/DataJsonEncoder.scala (hunk 3 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/serialization/DataJsonEncoder.scala:99]: see above

**Review by kushti — APPROVED** [review]
